### PR TITLE
prov/gni: Review and update use of GNIX_ERR #467

### PR DESCRIPTION
**Pull request by chuckfossen**
There are a few GNIX_ERR that I did not change that also need to be reviewed to verify it's the correct behavior.
@jswaro @sungeunchoi

### CODE REVIEW

File: prov/gni/include/gnix_util.h [hunk 1]
```diff
@@ -99,6 +99,8 @@ extern atomic_t gnix_debug_next_tid;
 #define GNIX_DEBUG(subsystem, ...)                                             \
 	GNIX_LOG_INTERNAL(FI_DBG, subsystem, __VA_ARGS__)
 #define GNIX_ERR(subsystem, ...)                                               \
+	GNIX_LOG_INTERNAL(GNIX_FI_PRINT, subsystem, __VA_ARGS__)
```
Comment (sungeunchoi) [inline prov/gni/include/gnix_util.h:102]: I think GNIX_ERR should tack on some extra identifier or something in the string it prints out, so we can distinguish it from GNIX_WARN.  I think the best way to do that is to add something to GNIX_LOG_INTERNAL.

**Comment by sungeunchoi** [comment]
Should the GNIX_ERR calls become GNIX_ABORT?  When @jswaro added the abort to GNIX_ERR, I thought we decided on the ones that should abort and turn them into GNIX_ERR calls.

**Comment by chuckfossen** [comment]
@sungeunchoi, I updated issue #467 with comments on my approach. Did you notice it?

File: prov/gni/include/gnix_util.h (hunk 1 above)
Comment (chuckfossen) [inline prov/gni/include/gnix_util.h:102]: I'll look into this.

**Comment by sungeunchoi** [comment]
Sorry I saw that after I posted my comment.  Can you point to the ones that you are not sure of?

**Comment by chuckfossen** [comment]
A grep of the ones not changed:

./src/gnix_mr.c:398:                GNIX_ERR(FI_LOG_MR,
./src/gnix_mr.c:411:                GNIX_ERR(FI_LOG_MR,
./src/gnix_mr.c:459:                        GNIX_ERR(FI_LOG_MR,
./src/gnix_mr.c:480:                        GNIX_ERR(FI_LOG_MR,
./src/gnix_mr.c:492:                    GNIX_ERR(FI_LOG_MR,
./src/gnix_mr.c:913:            GNIX_ERR(FI_LOG_MR,
./src/gnix_mr.c:922:            GNIX_ERR(FI_LOG_MR,
./src/gnix_mr.c:944:            GNIX_ERR(FI_LOG_MR,
./src/gnix_mr.c:1096:       GNIX_ERR(FI_LOG_MR, "failure in registration, can't really "
./src/gnix_mr.c:1182:               GNIX_ERR(FI_LOG_MR,
./src/gnix_mr.c:1287:       GNIX_ERR(FI_LOG_MR, "failed to insert registration "
./src/gnix_nic.c:559:       GNIX_ERR(FI_LOG_EP_CTRL, "No CDM attached to nic, nic=%p");
./src/gnix_wait.c:153:          GNIX_ERR(WAIT_SUB, "failed to signal wait object.\n");
./src/gnix_wait.c:159:      GNIX_ERR(WAIT_SUB,

List of changes with some context:
/Users/chuckf/work/libfabrics/libfabric-cray/prov/gni/src/gnix_mr.c:
  396           grc = __mr_cache_entry_put(cache, parent);
  397           if (unlikely(grc != GNI_RC_SUCCESS)) {
  398:              GNIX_ERR(FI_LOG_MR,
  399                       "failed to release reference to parent, "
  400                       "parent=%p refs=%d\n",
  ...
  409           iter = rbtFind(cache->inuse.rb_tree, &entry->key);
  410           if (unlikely(!iter)) {
  411:              GNIX_ERR(FI_LOG_MR,
  412                       "failed to find entry in the inuse cache\n");
  413           } else {
  ...
  457                   grc = __mr_cache_entry_destroy(c_entry);
  458                   if (grc != GNI_RC_SUCCESS) {
  459:                      GNIX_ERR(FI_LOG_MR,
  460                               "failed to destroy a "
  461                               "registration, entry=%p grc=%d\n",
  ...
  478                   grc = __mr_cache_entry_destroy(entry);
  479                   if (grc != GNI_RC_SUCCESS) {
  480:                      GNIX_ERR(FI_LOG_MR,
  481                               "failed to destroy a "
  482                               "registration, entry=%p grc=%d\n",
  ...
  490                       entry);
  491               if (rc != RBT_STATUS_OK) {
  492:                  GNIX_ERR(FI_LOG_MR,
  493                           "could not insert into stale rb tree,"
  494                           " rc=%d key.address=%llu key.length=%llu entry=%p",
  ...
  911       rc = __mr_cache_lru_dequeue(cache, &entry);
  912       if (unlikely(rc != FI_SUCCESS)) {
  913:          GNIX_ERR(FI_LOG_MR,
  914                   "list may be corrupt, no entries from lru pop\n");
  915           break;
  ...
  920       iter = rbtFind(cache->stale.rb_tree, &entry->key);
  921       if (unlikely(!iter)) {
  922:          GNIX_ERR(FI_LOG_MR,
  923                   "lru entries MUST be present in the cache,"
  924                   " could not find key in stale tree\n");
  ...
  942       rc = rbtErase(cache->stale.rb_tree, iter);
  943       if (unlikely(rc != RBT_STATUS_OK)) {
  944:          GNIX_ERR(FI_LOG_MR,
  945                   "failed to erase lru entry from stale tree\n");
  946           break;
  ...
 1094           vmdh_index, entry, &new_key);
 1095   if (ret) {
 1096:      GNIX_ERR(FI_LOG_MR, "failure in registration, can't really "
 1097               "recover at this point since we've already retired "
 1098               "the existing entries\n");
 ....
 1180           rc = rbtErase(cache->stale.rb_tree, iter);
 1181           if (unlikely(rc != RBT_STATUS_OK)) {
 1182:              GNIX_ERR(FI_LOG_MR,
 1183                       "failed to erase entry from stale tree\n");
 1184           } else {
 ....
 1285           current_entry);
 1286   if (unlikely(rc != RBT_STATUS_OK)) {
 1287:      GNIX_ERR(FI_LOG_MR, "failed to insert registration "
 1288               "into cache, ret=%i\n", rc);
 1289  

/Users/chuckf/work/libfabrics/libfabric-cray/prov/gni/src/gnix_nic.c:
  557  
  558   if (!nic->gni_cdm_hndl) {
  559:      GNIX_ERR(FI_LOG_EP_CTRL, "No CDM attached to nic, nic=%p");
  560   }
  561  

/Users/chuckf/work/libfabrics/libfabric-cray/prov/gni/src/gnix_wait.c:
  151   case FI_WAIT_FD:
  152       if (write(wait_priv->fd[WAIT_WRITE], &msg, len) != len)
  153:          GNIX_ERR(WAIT_SUB, "failed to signal wait object.\n");
  154       break;
  155   case FI_WAIT_MUTEX_COND:
  ...
  157       break;
  158   default:
  159:      GNIX_ERR(WAIT_SUB,
  160            "error signaling wait object: type: %d not supported.\n",
  161            wait_priv->type);

**Comment by sungeunchoi** [comment]
@jswaro @hppritcha please review Chuck's list of GNIX_ERR above and comment on whether they should abort or not.

**Comment by hppritcha** [comment]
I will review the list more closely but in general do not want to invoke abort.

**Comment by chuckfossen** [comment]
To be clear, the approaching I took here is:
- Add a GNIX_FATAL: prints message unconditionally and aborts
- GNIX_ERR: prints message unconditionally
- GNIX_WARN: no change
- Change GNIX_ERR calls to GNIX_WARN where the error is being handled appropriately.

**Comment by hppritcha** [comment]
I'll let @jswaro speak for gnix_mr.c, but I think the `GNIX_ERR` in gnix_nic.c should be replaced with `GNIX_WARN`.  I think the first `GNIX_ERR` at line 153 in gnix_wait.c should be replaced with `GNIX_WARN`.  I'm actually not sure its possible to get to line 159.  I think an error would have happened earlier when trying to create a waitset.  I'd suggest replacing that one with `GNIX_WARN` as well.

**Comment by jswaro** [comment]
I am fine with not changing the ones in gnix_mr.c. The only time I throw an error is when something happened that did not match my assertions.

**Comment by jswaro** [comment]
:+1:

**Pull request merged**